src/opencl/mod.rs

```diff
@@ -201,74 +178,100 @@ fn get_device_by_index(index: usize) -> Option<&'static Device> {
 }
 
 pub struct Program {
-    device: Device,
-    program: ocl::Program,
-    queue: ocl::Queue,
+    device_name: String,
+    program: opencl3::program::Program,
+    queue: opencl3::command_queue::CommandQueue,
+    context: opencl3::context::Context,
+    kernels: HashMap<String, opencl3::kernel::Kernel>,
 }
 
 impl Program {
-    pub fn device(&self) -> Device {
-        self.device.clone()
+    pub fn device_name(&self) -> &str {
+        &self.device_name
     }
-    pub fn from_opencl(device: Device, src: &str) -> GPUResult<Program> {
-        let cached = utils::cache_path(&device, src)?;
+    pub fn from_opencl(device: &Device, src: &str) -> GPUResult<Program> {
+        let cached = utils::cache_path(device, src)?;
         if std::path::Path::exists(&cached) {
             let bin = std::fs::read(cached)?;
             Program::from_binary(device, bin)
         } else {
-            let context = ocl::Context::builder()
-                .platform(device.platform)
-                .devices(device.device)
-                .build()?;
-            let program = ocl::Program::builder()
-                .src(src)
-                .devices(ocl::builders::DeviceSpecifier::Single(device.device))
-                .build(&context)?;
-            let queue = ocl::Queue::new(&context, device.device, None)?;
+            let context = opencl3::context::Context::from_device(&device.device)?;
+            let mut program = opencl3::program::Program::create_from_source(&context, src)?;
+            if let Err(build_error) = program.build(context.devices(), "") {
+                let log = program.get_build_log(context.devices()[0])?;
```

Comment:
is this guranteed to always have a device? otherwise this could panic

Reply:
Yes, it is, as the context was created from a device.